File: Microsoft.CodeAnalysis.VisualBasic/Microsoft.CodeAnalysis.VisualBasic.Syntax/TypeBlockSyntax.cs
using System;
using System.ComponentModel;

namespace Microsoft.CodeAnalysis.VisualBasic.Syntax
{
	public abstract class TypeBlockSyntax : DeclarationStatementSyntax
	{
		internal SyntaxNode _inherits;

		internal SyntaxNode _implements;

		internal SyntaxNode _members;

		public SyntaxList<InheritsStatementSyntax> Inherits => GetInheritsCore();

		public SyntaxList<ImplementsStatementSyntax> Implements => GetImplementsCore();

		public SyntaxList<StatementSyntax> Members => GetMembersCore();

		public abstract TypeStatementSyntax BlockStatement { get; }

		public abstract EndBlockStatementSyntax EndBlockStatement { get; }

		[EditorBrowsable(EditorBrowsableState.Never)]
		[Obsolete("This member is obsolete. Use BlockStatement or a more specific property (e.g. ClassStatement) instead.", true)]
		public TypeStatementSyntax Begin => BlockStatement;

		[EditorBrowsable(EditorBrowsableState.Never)]
		[Obsolete("This member is obsolete. Use EndBlockStatement or a more specific property (e.g. EndClassStatement) instead.", true)]
		public EndBlockStatementSyntax End => EndBlockStatement;

		internal TypeBlockSyntax(GreenNode green, SyntaxNode parent, int startLocation)
			: base(green, parent, startLocation)
		{
		}

		internal virtual SyntaxList<InheritsStatementSyntax> GetInheritsCore()
		{
			SyntaxNode redAtZero = GetRedAtZero(ref _inherits);
			return new SyntaxList<InheritsStatementSyntax>(redAtZero);
		}

		public TypeBlockSyntax WithInherits(SyntaxList<InheritsStatementSyntax> inherits)
		{
			return WithInheritsCore(inherits);
		}

		internal abstract TypeBlockSyntax WithInheritsCore(SyntaxList<InheritsStatementSyntax> inherits);

		public TypeBlockSyntax AddInherits(params InheritsStatementSyntax[] items)
		{
			return AddInheritsCore(items);
		}

		internal abstract TypeBlockSyntax AddInheritsCore(params InheritsStatementSyntax[] items);

		internal virtual SyntaxList<ImplementsStatementSyntax> GetImplementsCore()
		{
			SyntaxNode red = GetRed(ref _implements, 1);
			return new SyntaxList<ImplementsStatementSyntax>(red);
		}

		public TypeBlockSyntax WithImplements(SyntaxList<ImplementsStatementSyntax> implements)
		{
			return WithImplementsCore(implements);
		}

		internal abstract TypeBlockSyntax WithImplementsCore(SyntaxList<ImplementsStatementSyntax> implements);

		public TypeBlockSyntax AddImplements(params ImplementsStatementSyntax[] items)
		{
			return AddImplementsCore(items);
		}

		internal abstract TypeBlockSyntax AddImplementsCore(params ImplementsStatementSyntax[] items);

		internal virtual SyntaxList<StatementSyntax> GetMembersCore()
		{
			SyntaxNode red = GetRed(ref _members, 2);
			return new SyntaxList<StatementSyntax>(red);
		}

		public TypeBlockSyntax WithMembers(SyntaxList<StatementSyntax> members)
		{
			return WithMembersCore(members);
		}

		internal abstract TypeBlockSyntax WithMembersCore(SyntaxList<StatementSyntax> members);

		public TypeBlockSyntax AddMembers(params StatementSyntax[] items)
		{
			return AddMembersCore(items);
		}

		internal abstract TypeBlockSyntax AddMembersCore(params StatementSyntax[] items);

		public abstract TypeBlockSyntax WithBlockStatement(TypeStatementSyntax blockStatement);

		public abstract TypeBlockSyntax WithEndBlockStatement(EndBlockStatementSyntax endBlockStatement);

		[EditorBrowsable(EditorBrowsableState.Never)]
		[Obsolete("This member is obsolete. Use WithBlockStatement or a more specific property (e.g. WithClassStatement) instead.", true)]
		public TypeBlockSyntax WithBegin(TypeStatementSyntax begin)
		{
			return WithBlockStatement(begin);
		}

		[EditorBrowsable(EditorBrowsableState.Never)]
		[Obsolete("This member is obsolete. Use WithEndBlockStatement or a more specific property (e.g. WithEndClassStatement) instead.", true)]
		public TypeBlockSyntax WithEnd(EndBlockStatementSyntax end)
		{
			return WithEndBlockStatement(end);
		}
	}
}

File: Microsoft.CodeAnalysis.VisualBasic/Microsoft.CodeAnalysis/VisualBasic/Syntax/TypeBlockSyntax.vb
Imports Microsoft.CodeAnalysis
Imports System
Imports System.ComponentModel

Namespace Microsoft.CodeAnalysis.VisualBasic.Syntax
	Public MustInherit Class TypeBlockSyntax
		Inherits DeclarationStatementSyntax
		Friend _inherits As SyntaxNode

		Friend _implements As SyntaxNode

		Friend _members As SyntaxNode

		<EditorBrowsable(EditorBrowsableState.Never)>
		<Obsolete("This member is obsolete. Use BlockStatement or a more specific property (e.g. ClassStatement) instead.", True)>
		Public ReadOnly Property Begin As TypeStatementSyntax
			Get
				Return Me.BlockStatement
			End Get
		End Property

		Public MustOverride ReadOnly Property BlockStatement As TypeStatementSyntax

		<EditorBrowsable(EditorBrowsableState.Never)>
		<Obsolete("This member is obsolete. Use EndBlockStatement or a more specific property (e.g. EndClassStatement) instead.", True)>
		Public ReadOnly Property [End] As EndBlockStatementSyntax
			Get
				Return Me.EndBlockStatement
			End Get
		End Property

		Public MustOverride ReadOnly Property EndBlockStatement As EndBlockStatementSyntax

		Public ReadOnly Property [Implements] As SyntaxList(Of ImplementsStatementSyntax)
			Get
				Return Me.GetImplementsCore()
			End Get
		End Property

		Public ReadOnly Property [Inherits] As SyntaxList(Of InheritsStatementSyntax)
			Get
				Return Me.GetInheritsCore()
			End Get
		End Property

		Public ReadOnly Property Members As SyntaxList(Of StatementSyntax)
			Get
				Return Me.GetMembersCore()
			End Get
		End Property

		Friend Sub New(ByVal green As GreenNode, ByVal parent As SyntaxNode, ByVal startLocation As Integer)
			MyBase.New(green, parent, startLocation)
		End Sub

		Public Function AddImplements(ByVal ParamArray items As ImplementsStatementSyntax()) As TypeBlockSyntax
			Return Me.AddImplementsCore(items)
		End Function

		Friend MustOverride Function AddImplementsCore(ByVal ParamArray items As ImplementsStatementSyntax()) As TypeBlockSyntax

		Public Function AddInherits(ByVal ParamArray items As InheritsStatementSyntax()) As TypeBlockSyntax
			Return Me.AddInheritsCore(items)
		End Function

		Friend MustOverride Function AddInheritsCore(ByVal ParamArray items As InheritsStatementSyntax()) As TypeBlockSyntax

		Public Function AddMembers(ByVal ParamArray items As StatementSyntax()) As TypeBlockSyntax
			Return Me.AddMembersCore(items)
		End Function

		Friend MustOverride Function AddMembersCore(ByVal ParamArray items As StatementSyntax()) As TypeBlockSyntax

		Friend Overridable Function GetImplementsCore() As SyntaxList(Of ImplementsStatementSyntax)
			Return New SyntaxList(Of ImplementsStatementSyntax)(MyBase.GetRed(Me._implements, 1))
		End Function

		Friend Overridable Function GetInheritsCore() As SyntaxList(Of InheritsStatementSyntax)
			Return New SyntaxList(Of InheritsStatementSyntax)(MyBase.GetRedAtZero(Me._inherits))
		End Function

		Friend Overridable Function GetMembersCore() As SyntaxList(Of StatementSyntax)
			Return New SyntaxList(Of StatementSyntax)(MyBase.GetRed(Me._members, 2))
		End Function

		<EditorBrowsable(EditorBrowsableState.Never)>
		<Obsolete("This member is obsolete. Use WithBlockStatement or a more specific property (e.g. WithClassStatement) instead.", True)>
		Public Function WithBegin(ByVal begin As TypeStatementSyntax) As TypeBlockSyntax
			Return Me.WithBlockStatement(begin)
		End Function

		Public MustOverride Function WithBlockStatement(ByVal blockStatement As TypeStatementSyntax) As TypeBlockSyntax

		<EditorBrowsable(EditorBrowsableState.Never)>
		<Obsolete("This member is obsolete. Use WithEndBlockStatement or a more specific property (e.g. WithEndClassStatement) instead.", True)>
		Public Function WithEnd(ByVal [end] As EndBlockStatementSyntax) As TypeBlockSyntax
			Return Me.WithEndBlockStatement([end])
		End Function

		Public MustOverride Function WithEndBlockStatement(ByVal endBlockStatement As EndBlockStatementSyntax) As TypeBlockSyntax

		Public Function WithImplements(ByVal [implements] As SyntaxList(Of ImplementsStatementSyntax)) As TypeBlockSyntax
			Return Me.WithImplementsCore([implements])
		End Function

		Friend MustOverride Function WithImplementsCore(ByVal [implements] As SyntaxList(Of ImplementsStatementSyntax)) As TypeBlockSyntax

		Public Function WithInherits(ByVal [inherits] As SyntaxList(Of InheritsStatementSyntax)) As TypeBlockSyntax
			Return Me.WithInheritsCore([inherits])
		End Function

		Friend MustOverride Function WithInheritsCore(ByVal [inherits] As SyntaxList(Of InheritsStatementSyntax)) As TypeBlockSyntax

		Public Function WithMembers(ByVal members As SyntaxList(Of StatementSyntax)) As TypeBlockSyntax
			Return Me.WithMembersCore(members)
		End Function

		Friend MustOverride Function WithMembersCore(ByVal members As SyntaxList(Of StatementSyntax)) As TypeBlockSyntax
	End Class
End Namespace